File: dotnet/Base/GameX.Base/Store_Steam.cs
using Microsoft.Win32;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;

namespace GameX.Store
{
    /// <summary>
    /// Store_Steam
    /// </summary>
    internal static class Store_Steam
    {
        internal static Dictionary<string, string> SteamPaths = new Dictionary<string, string>();

        static Store_Steam()
        {
            // get dbPath
            var root = GetPath();
            if (root == null) return;

            // query games
            var libraryFolders = AcfStruct.Read(Path.Join(root, "steamapps", "libraryfolders.vdf"));
            foreach (var folder in libraryFolders.Get["libraryfolders"].Get.Values)
            {
                var path = folder.Value["path"];
                if (!Directory.Exists(path)) continue;
                foreach (var appId in folder.Get["apps"].Value.Keys)
                {
                    var appManifest = AcfStruct.Read(Path.Join(path, "steamapps", $"appmanifest_{appId}.acf"));
                    if (appManifest == null) continue;
                    // add appPath if exists
                    var appPath = Path.Join(path, "steamapps", Path.Join("common", appManifest.Get["AppState"].Value["installdir"]));
                    if (Directory.Exists(appPath)) SteamPaths.Add(appId, appPath);
                }
            }
        }

        static string GetPath()
        {
            IEnumerable<string> paths;
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                // windows paths
                var key = Registry.CurrentUser.OpenSubKey(@"SOFTWARE\Valve\Steam")
                    ?? RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64).OpenSubKey(@"SOFTWARE\Valve\Steam");
                if (key == null) return null;
                return (string)key.GetValue("SteamPath");
            }
            else if (RuntimeInformation.OSDescription.StartsWith("android-")) return null;
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                // linux paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                var search = new[] { ".steam", ".steam/steam", ".steam/root", ".local/share/Steam" };
                paths = search.Select(path => Path.Join(home, path, "appcache"));
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                // mac paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                var search = new[] { "Library/Application Support/Steam" };
                paths = search.Select(path => Path.Join(home, path, "appcache"));
            }
            else throw new PlatformNotSupportedException();
            return paths.FirstOrDefault(Directory.Exists);
        }

        #region steamapp

        public class AcfStruct
        {
            public static AcfStruct Read(string path) => File.Exists(path) ? new AcfStruct(File.ReadAllText(path)) : null;
            public Dictionary<string, AcfStruct> Get = new Dictionary<string, AcfStruct>();
            public Dictionary<string, string> Value = new Dictionary<string, string>();

            public AcfStruct(string region)
            {
                int lengthOfRegion = region.Length, index = 0;
                while (lengthOfRegion > index)
                {
                    var firstStart = region.IndexOf('"', index);
                    if (firstStart == -1) break;
                    var firstEnd = region.IndexOf('"', firstStart + 1);
                    index = firstEnd + 1;
                    var first = region.Substring(firstStart + 1, firstEnd - firstStart - 1);
                    int secondStart = region.IndexOf('"', index), secondOpen = region.IndexOf('{', index);
                    if (secondStart == -1)
                        Get.Add(first, null);
                    else if (secondOpen == -1 || secondStart < secondOpen)
                    {
                        var secondEnd = region.IndexOf('"', secondStart + 1);
                        index = secondEnd + 1;
                        var second = region.Substring(secondStart + 1, secondEnd - secondStart - 1);
                        Value.Add(first, second.Replace(@"\\", @"\"));
                    }
                    else
                    {
                        var secondClose = NextEndOf(region, '{', '}', secondOpen + 1);
                        var acfs = new AcfStruct(region.Substring(secondOpen + 1, secondClose - secondOpen - 1));
                        index = secondClose + 1;
                        Get.Add(first, acfs);
                    }
                }
            }

            static int NextEndOf(string str, char open, char close, int startIndex)
            {
                if (open == close) throw new Exception("\"Open\" and \"Close\" char are equivalent!");
                int openItem = 0, closeItem = 0;
                for (var i = startIndex; i < str.Length; i++)
                {
                    if (str[i] == open) openItem++;
                    if (str[i] == close) { closeItem++; if (closeItem > openItem) return i; }
                }
                throw new Exception("Not enough closing characters!");
            }

            public override string ToString() => ToString(0);
            public string ToString(int depth)
            {
                var b = new StringBuilder();
                foreach (var item in Value)
                {
                    b.Append('\t', depth);
                    b.AppendFormat("\"{0}\"\t\t\"{1}\"\r\n", item.Key, item.Value);
                }
                foreach (var item in Get)
                {
                    b.Append('\t', depth);
                    b.AppendFormat("\"{0}\"\n", item.Key);
                    b.Append('\t', depth);
                    b.AppendLine("{");
                    b.Append(item.Value.ToString(depth + 1));
                    b.Append('\t', depth);
                    b.AppendLine("}");
                }
                return b.ToString();
            }
        }

        #endregion
    }
}

File: Python/gamex/store_steam.py
import os, platform, winreg

# AcfStruct
class AcfStruct:
    def read(path):
        if not os.path.exists(path): return None
        with open(path, 'r') as f: return AcfStruct(f.read())

    def __init__(self, region):
        def nextEndOf(str, open, close, startIndex):
            if open == close:
                raise Exception('"Open" and "Close" char are equivalent!')
            openItem = 0; closeItem = 0
            for i in range(startIndex, len(str)):
                if str[i] == open: openItem += 1
                if str[i] == close:
                    closeItem += 1
                    if closeItem > openItem: return i
            raise Exception('Not enough closing characters!')
        self.get = {}
        self.value = {}
        lengthOfRegion = len(region); index = 0
        while (lengthOfRegion > index):
            firstStart = region.find('"', index)
            if firstStart == -1: break
            firstEnd = region.find('"', firstStart + 1)
            index = firstEnd + 1
            first = region[firstStart + 1:firstEnd]
            secondStart = region.find('"', index)
            secondOpen = region.find('{', index)
            if secondStart == -1:
                self.get[first] = None
            elif secondOpen == -1 or secondStart < secondOpen:
                secondEnd = region.find('"', secondStart + 1)
                index = secondEnd + 1
                second = region[secondStart + 1:secondEnd]
                self.value[first] = second.replace('\\\\', '\\')
            else:
                secondClose = nextEndOf(region, '{', '}', secondOpen + 1)
                acfs = AcfStruct(region[secondOpen + 1:secondClose])
                index = secondClose + 1
                self.get[first] = acfs

    def repr(self, depth):
        b = []
        for (k,v) in self.value.items():
            b.append(f'{"  "*depth}"{k}": "{v}"\n')
        for (k,v) in self.get.items():
            b.append(f'{"  "*depth}"{k}" {{\n{"  "*depth}')
            if not v is None: b.append(v.repr(depth + 1))
            b.append(f'}}\n')
        return ''.join(b)
    def __repr__(self): return self.repr(0)

def getPath():
    system = platform.system()
    if system == 'Windows':
        # windows paths
        try: key = winreg.OpenKey(winreg.HKEY_CURRENT_USER, 'SOFTWARE\\Valve\\Steam', 0, winreg.KEY_READ)
        except FileNotFoundError:
            try: key = winreg.OpenKey(winreg.HKEY_LOCAL_MACHINE, 'SOFTWARE\\Valve\\Steam', 0, winreg.KEY_READ | winreg.KEY_WOW64_32KEY)
            except FileNotFoundError: return None
        return winreg.QueryValueEx(key, 'SteamPath')[0]
    elif system == 'Linux':
        # linux paths
        home = os.path.expanduser('~')
        search = ['.steam', '.steam/steam', '.steam/root', '.local/share/Steam']
        paths = [os.path.join(home, path, 'appcache') for path in search]
    elif system == 'Darwin':
        # mac paths
        home = '/Users/Shared'
        search = ['Library/Application Support/Steam']
        paths = [os.path.join(home, path, 'appcache') for path in search]
    else: raise Exception(f'Unknown platform: {system}')
    return next(iter(x for x in paths if os.path.isdir(x)), None)
    
# get steamPaths
steamPaths = {}
root = getPath()
if root:
    # query games
    libraryFolders = AcfStruct.read(os.path.join(root, 'steamapps', 'libraryfolders.vdf'))
    for folder in libraryFolders.get['libraryfolders'].get.values():
        path = folder.value['path']
        if not os.path.isdir(path): continue
        for appId in folder.get['apps'].value.keys():
            appManifest = AcfStruct.read(os.path.join(path, 'steamapps', f'appmanifest_{appId}.acf'))
            if appManifest is None: continue
            # add appPath if exists
            appPath = os.path.join(path, 'steamapps', 'common', appManifest.get['AppState'].value['installdir'])
            if os.path.isdir(appPath): steamPaths[appId] = appPath

# print(steamPaths)
